### ISSUE TITLE
GRPO Trainer cannot use rewards as best metric to save best model

### ISSUE THREAD

**Issue by lohzhunyewcs** [huggingface/trl#3384]
### Reproduction

Hi currently I want to use reward as the best metric but it seems like it is not an option as when I tried to use it the trainer outputs an error that I can only use eval_loss, rutime and samples_per_second. Is there anyway to specify it? Also is it possible to use train metrics instead of just eval metrics(besides just setting train set as eval set)

Reproducible steps under GRPO config add
```
metrics_for_best_model="eval_reward"
```

### System Info

trl==0.16.1
torch==2.6.0
transformers==4.51.1
python==3.12.9

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by andriyrizhiy** [comment]
Same issue with 
transformers==4.55.1
trl==0.21.0

Does anyone have progress with it?

**Comment by LeonMalteW** [comment]
try this workaround

```
from transformers.trainer_utils import SaveStrategy
from trl import GRPOTrainer
import numpy as np

class GRPOTrainerEval(GRPOTrainer):

    def _determine_best_metric(self, metrics, trial):
        """
        Determine if the model should be saved based on the evaluation metrics.

        Returns:
            bool: True if a new best metric was found, else False
        """
        is_new_best_metric = False

        if self.args.metric_for_best_model is not None:
            metric_to_check = self.args.metric_for_best_model

            if not metric_to_check.startswith("eval_"):
                metric_to_check = f"eval_{metric_to_check}"

            try:
                # self._metrics["train"] is the only metric that is used to confirm you could jsut print 
                # self._metrics["eval"] but in my runs this is always empty 

                metric_value = self._metrics["train"][metric_to_check[5:]] # [5:] removes the "eval_"
                if metric_value is None:
                    metric_value = metrics[metric_to_check]
            except KeyError as exc:
                raise KeyError(
                    f"The `metric_for_best_model` training argument is set to '{metric_to_check}', which is not found in the evaluation metrics. "
                    f"The available evaluation metrics are: {list(metrics.keys())}. Consider changing the `metric_for_best_model` via the TrainingArguments."
                ) from exc

            operator = np.greater if self.args.greater_is_better else np.less

            if self.state.best_metric is None:
                self.state.best_metric = float("-inf") if self.args.greater_is_better else float("inf")

            if operator(metric_value, self.state.best_metric):
                self.state.best_metric = metric_value

                if self.args.save_strategy in [SaveStrategy.STEPS, SaveStrategy.EPOCH]:
                    self.state.best_global_step = self.state.global_step

                is_new_best_metric = True

        return is_new_best_metric

```

I don't understand the idea behind the custom metric, but the whole thing is too complicated for me to provide an actual fix for this.

I don't know what this 'fix' breaks, this could total screw with some things!

**Comment by kevinlu1248** [comment]
Bump on this, still running into this issue.

**Comment by francescotaioli** [comment]
Me too

**Comment by practicingman** [comment]
Me too 
trl==0.26.2
transformers==4.57.6

**Comment by prd-tuong-nguyen** [comment]
any update?

**Comment by prd-tuong-nguyen** [comment]
@LeonMalteW Thank you for the workaround, it seems that you are getting the reward of the training set.
Using your approach but change `metric_value = self._metrics["train"][metric_to_check[5:]]` to `metric_value = self.state.log_history[-1][metric_to_check]` works well on my side

**Comment by albertvillanova** [comment]
Hey @lohzhunyewcs, sorry for the very slow reply — we're catching up on the backlog. 

This was a genuine bug: before v1.5.0, GRPO's `log()` rebuilt the log dict (`logs = {**logs, **metrics}`) instead of updating it in place, so trainer-specific eval metrics like `eval_reward` never reached transformers' best-model check — which is exactly why you saw the KeyError listing only `eval_loss/eval_runtime/eval_samples_per_second`. It was fixed in #5811 (switched to an in-place `logs.update(...)`, see https://github.com/huggingface/trl/blob/e1b2e21975994e676d9a676e3ab005e9bee38e40/trl/trainer/grpo_trainer.py#L3305) and shipped in v1.5.0. On a current version, `metric_for_best_model="eval_reward"` together with `eval_strategy="steps"`, an `eval_dataset`, `greater_is_better=True`, and `load_best_model_at_end=True` should work. 

On your second question: best-model selection in `transformers` is tied to eval metrics (it always looks for an `eval_`-prefixed key after an eval pass), so there's no native "save-best on the training reward" — passing a small eval split is the intended path. 

Closing as fixed, but happy to reopen if you still hit this on >= v1.5.0.
